parsers/Parser.js

```diff
@@ -7,31 +9,97 @@ module.exports = class Parser {
-        }
+    get ast() {
+        throw new Error('ast getter must be implemented in the child class.');
+    }
 
-        return this.flagDuplicates(rules);
+    getNodeKey(node) {
+        // This default implementation fits for YAML, JSON, and Package parsers.
+        // JS parser overrides it.
+        return node.key.value;
+    }
+
+    /** Gets properties of an object-like AST node. */
+    getProps(node) {
+        throw new Error('getProps method must be implemented in the child class.');
+    }
+
+    /** Gets elements of an array-like AST node. */
+    getItems(node) {
+        throw new Error('getItems method must be implemented in the child class.');
+    }
+
+    /** Flatmaps `callback` over `node`'s child property (or children properties) named `key`. */
+    mapChildren(node, key, callback) {
+        return this.getProps(node)
+            .filter(prop => this.getNodeKey(prop) === key)
+            .flatMap(prop => callback(prop.value));
+    }
+
+    /** Gets both plain and overridden rules from `node`. */
+    getAllRules(node) {
+        const mainRules = this.getRulesFromNode(node);
+        const overrideRules = this.getOverridesFromNode(node);
+
+        return mainRules.concat(overrideRules);
+    }
+
+    /** Gets rules from AST node, assuming this node has 'rules' child. */
+    getRulesFromNode(node) {
+        return this.mapChildren(node, 'rules', value =>
+            this.getProps(value).map(rule => this.getRule(rule)),
+        );
+    }
+
+    /** Gets rules from AST node, assuming this node has 'overrides' child. */
+    getOverridesFromNode(node) {
+        return this.mapChildren(node, 'overrides', value =>
+            this.getItems(value).flatMap(item => this.getRulesFromNode(item)),
+        );
+    }
+
+    /** Gets start and end positions of the rule location in the document. */
+    getRuleLocation(rule) {
+        throw new Error('getRuleLocation method must be implemented in the child class.');
+    }
+
+    getRule(rule) {
+        const [start, end] = this.getRuleLocation(rule);
+
+        const name = this.getNodeKey(rule);
+        const keyRange = this.document.validateRange(new vscode.Range(start, end));
+        const eol = new vscode.Position(keyRange.start.line, Number.MAX_SAFE_INTEGER);
+        const eolRange = this.document.validateRange(new vscode.Range(eol, eol));
+
+        return { name, keyRange, eolRange };
     }
 
     flagDuplicates(rules) {
-        let ruleNames = rules.map(rule => rule.name);
+        const ruleNames = rules.map(rule => rule.name);
 
-        return rules.map((rule, index) => {
-            let otherIndex = ruleNames.indexOf(rule.name, index + 1);
-            if (otherIndex > -1) {
-                rule.duplicate = true;
+        rules.forEach((rule, index) => {
+            if (rule.duplicate) {
+                return;
+            }
+            const otherIndex = ruleNames.indexOf(rule.name, index + 1);
+            if (~otherIndex) {
```

Comment:
I would prefer to avoid clever tricks such as this.  It lowers code readability.  Please change this to:

```
if (otherIndex > -1) {
```